# scDeconvolution_course
Single-cell deconvolution methods course

June 28th 2024  

-Datasets and associated metadata: PDAC TCGA (downsampled)

-Inference of cell-types using BayesPrism, Bisque, MuSiC, CIBERSORTx, and EPIC algorithms  


Resources/Vignettes:  

Bisque vignette: http://127.0.0.1:27074/library/BisqueRNA/doc/bisque.html  
BayesPrism tutorial: https://www.bayesprism.org/pages/tutorial_deconvolution  
MuSic tutorial: https://xuranw.github.io/MuSiC/articles/MuSiC.html  
Omnideconv mouse https://omnideconv.org/immunedeconv/articles/detailed_example_mouse.html  

-Avila Cobos, F., Alquicira-Hernandez, J., Powell, J.E. et al. Benchmarking of cell type deconvolution pipelines for transcriptomics data. Nat Commun 11, 5650 (2020). https://doi.org/10.1038/s41467-020-19015-1  

-Huuki-Myers, L. A., Montgomery, K. D., Kwon, S. H., Cinquemani, S., Eagles, N. J., Gonzalez-Padilla, D., Maden, S. K., Kleinman, J. E., Hyde, T. M., Hicks, S. C., Maynard, K. R., & Collado-Torres, L. (2024). Benchmark of cellular deconvolution methods using a multi-assay reference dataset from postmortem human prefrontal cortex. bioRxiv : the preprint server for biology, 2024.02.09.579665. https://doi.org/10.1101/2024.02.09.579665 

-Newman, A.M., Steen, C.B., Liu, C.L. et al. Determining cell type abundance and expression from bulk tissues with digital cytometry. Nat Biotechnol 37, 773â€“782 (2019). https://doi.org/10.1038/s41587-019-0114-2 https://www.nature.com/articles/s41587-019-0114-2  

-Jin, H., Liu, Z. A benchmark for RNA-seq deconvolution analysis under dynamic testing environments. Genome Biol 22, 102 (2021). https://doi.org/10.1186/s13059-021-02290-6 

-Chakravarthy, A., Furness, A., Joshi, K. et al. Pan-cancer deconvolution of tumour composition using DNA methylation. Nat Commun 9, 3220 (2018). https://doi.org/10.1038/s41467-018-05570-1 

-Trang Le, Rachel A Aronow, Arkadz Kirshtein, Leili Shahriyari, A review of digital cytometry methods: estimating the relative abundance of cell types in a bulk of cells, Briefings in Bioinformatics, Volume 22, Issue 4, July 2021, bbaa219, https://doi.org/10.1093/bib/bbaa219 

-Sutton, G.J., Poppe, D., Simmons, R.K. et al. Comprehensive evaluation of deconvolution methods for human brain gene expression. Nat Commun 13, 1358 (2022). https://doi.org/10.1038/s41467-022-28655-4 

-Tran, K.A., Addala, V., Johnston, R.L. et al. Performance of tumour microenvironment deconvolution methods in breast cancer using single-cell simulated bulk mixtures. Nat Commun 14, 5758 (2023). https://www.nature.com/articles/s41467-023-41385-5 

-Vathrakokoili Pournara A, Miao Z, Beker OY, Nolte N, Brazma A, Papatheodorou I. CATD: a reproducible pipeline for selecting cell-type deconvolution methods across tissues. Bioinform Adv. 2024 Mar 23;4(1):vbae048. doi: 10.1093/bioadv/vbae048. PMID: 38638280; PMCID: PMC11023940. (EMBL)

-Dietrich A., Metorro L., Pelz K., Eder B., Zackl C., Marini F., Sturm G., List M., Finotello F. (2024).
Benchmarking second-generation methods for cell-type deconvolution of transcriptomic data
bioRxiv doi: https://doi.org/10.1101/2024.06.10.598226

-Sutton, G.J., Poppe, D., Simmons, R.K. et al. Comprehensive evaluation of deconvolution methods for human brain gene expression. Nat Commun 13, 1358 (2022). https://doi.org/10.1038/s41467-022-28655-4 https://www.nature.com/articles/s41467-022-28655-4


