---
title: "Single-cell deconvolution methods"
output:
  html_document:
    theme: simplex
    toc: yes
    toc_float:
      collapsed: true
---

This vignette provides an overview of first and second generation methods for cell-type deconvolution using transcriptomic data. It provides consise examples, along with some considerations and exercises. We will be working with Pancreatic Ductal Adenocarcinoma (PDAC) data (downsampled) from the TCGA project.

```{r, message=FALSE}
#load libraries
suppressWarnings(library(BayesPrism))
library(Seurat)
library(MuSiC)
library(Biobase)
library(BisqueRNA)
library(immunedeconv)
library(plotly)
library(tidyverse)
library(ggthemes)
library(ggpubr)
library(RColorBrewer)
library(colorspace)
library(cowplot)
```

```{r}
#working directory
setwd("/home/training/Estef-Gabriel/deconvolution_data")

#Load data - PDAC TCGA downsampled
#1) bulk data raw counts
bulkdata <- readRDS('tcga_pdac_counts.rds')
bulkdata <- as.data.frame(t(bulkdata)) #Different than Bisque, Bayes expects features as columns (sample by gene)
dim(bulkdata)
head(rownames(bulkdata)) #samples as rownames
head(colnames(bulkdata)) #genes as colnames

#2) Load clinical data
clindata <- readRDS('tcga_pdac_clindata.rds')
#glimpse(clindata)

#load sc Seurat object
sobj <- readRDS('sc_ref_pdac_seurat.rds')
sobj
colnames(sobj)
nrow(sobj) #18.5k
Features(sobj)
Layers(sobj)
```

## BayesPrism
```{r}
#rename cell_meta columns
cell_meta <- cell_meta %>% dplyr::rename("cell.type.labels" = "cell_type",
                                         "cell.state.labels" = "cell_state") 
glimpse(cell_meta)

#If you have different levels of granularity for cell annotation, you can pass it. But, rare cell types or cell states might be challenging to estimate.
#As a start we can run at a higher level of cell annotation
cell.type.labels <- cell_meta$cell.type.labels
cell.state.labels <- cell_meta$cell.state.labels

cell.type.labels <- as.character(cell.type.labels)
cell.state.labels <- as.character(cell.state.labels)

#cell.type.labels will be a character vector of the same length as nrow(sc.dat) to denote the cell type of each cell in the reference
sort(table(cell.type.labels))
#cell.state.labels is a character vector of the same length as nrow(sc.dat) to denote the cell state of each cell in the reference
sort(table(cell.state.labels))

cell_annot <- as.data.frame(table(cbind.data.frame(cell.state.labels, cell.type.labels)))

#Extracting the counts from the seurat object
sc.dat <- GetAssayData(sobj, assay='RNA', slot='counts')
sc.dat <- t(sc.dat) #Here we transpose the matrix 

# In this step, we need to convert the sparse matrix to a dense matrix
sc.dat <- as.matrix(sc.dat) #heavy
colnames(sc.dat)

#Before starting the deconvolution, Bayes allows you to check for correlation between your cell types
#Higher correlation means that the cell types share similarities
#pairwise correlation matrix 
plot.cor.phi (input=sc.dat,
              input.labels=cell.type.labels,
              title="cell state correlation",
              pdf.prefix = "cor_matrix",
              cexRow=0.5, cexCol=0.5,
              margins=c(4,4)
)

plot.cor.phi (input=sc.dat,
              input.labels=cell.state.labels,
              title="cell state correlation",
              pdf.prefix = "cor_matrix_cell_states",
              cexRow=0.4, cexCol=0.4,
              margins=c(5,5)
)


#distribution of outlier genes from scRNA-seq reference
sc.stat <- plot.scRNA.outlier(
  input=sc.dat, #make sure the colnames are gene symbol or ENSMEBL ID
  cell.type.labels=cell.type.labels,
  species="hs", #currently only human(hs) and mouse(mm) annotations are supported
  return.raw=TRUE, #return the data used for plotting.
  pdf.prefix = "outlier_sc"
)
gc()

#visualize outlier genes from bulk RNA-seq
bk.stat <- plot.bulk.outlier(
  bulk.input=bulkdata,#make sure the colnames are gene symbol or ENSMEBL ID
  sc.input=sc.dat, #make sure the colnames are gene symbol or ENSMEBL ID
  cell.type.labels=cell.type.labels,
  species="hs", #currently only human(hs) and mouse(mm) annotations are supported
  return.raw=TRUE,
  pdf.prefix = "outlier_bulk" 
)
gc()

#In this step, we filter our matrix removing features that add noise to our data
sc.dat.filtered <- cleanup.genes(input=sc.dat,
                                 input.type="count.matrix",
                                 species="hs",
                                 gene.group=c( "Rb","Mrp","other_Rb","chrM","MALAT1","chrX","chrY") ,
                                 exp.cells=5)
gc()
dim(sc.dat.filtered)


#Correlation between the features present in our single-cell reference and bulk samples
plot.bulk.vs.sc(sc.input = sc.dat.filtered,
                bulk.input = bulkdata, 
                pdf.prefix = "concordance_sc_bulk"
)
gc()


#Subset the reference to retain only protein coding genes
sc.dat.filtered.pc <-  select.gene.type(sc.dat.filtered, gene.type = "protein_coding")
gc()


#Create the prism object
myPrism <- new.prism(
  reference=sc.dat.filtered.pc,
  mixture=bulkdata,
  input.type="count.matrix",
  cell.type.labels = cell.type.labels,
  cell.state.labels = cell.type.labels,
  key='Malignant',
  outlier.cut=0.01,
  outlier.fraction=0.1,
)
gc()


#Run BayesPrism deconvolution
#?run.prism 
bp.res <- run.prism(prism = myPrism, n.cores=10)
bp.res
slotNames(bp.res)

#extract cell type fractions
cell_fraction <- get.fraction(bp.res, which.theta = 'final', state.or.type = 'type')
cell_fraction_df <- as.data.frame(cell_fraction)
write_csv(cell_fraction_df, "cell_fraction_df.csv")

t_cell_fraction <- t(cell_fraction)
t_cell_fraction_df <- as.data.frame(t_cell_fraction)
write_csv(t_cell_fraction_df, "transposed_cell_fraction_df.csv")

#save results
save(bp.res, file = "bayesprism_results.rdata")

#exercise downstream analysis

```

## 2. MuSic
```{r}

```

## 3. Bisque
```{r}

# Running deconvolution with Bisque (remember that bisque expects features as rows and cells/samples as columns. If it is not, you just have to transpose)
# data: PDAC TCGA downsampled

# Working directory
setwd("/home/training/Estef-Gabriel/deconvolution_data")

# Load libraries
library(Biobase)
library(BisqueRNA)
library(Seurat)

# Load bulk data
bulkdata <- readRDS("tcga_pdac_counts.rds")
bulkdata <- as.matrix(bulkdata)

# Convert bulk RNA-seq from matrix to an ExpressionSet object
bulk.eset <- ExpressionSet(assayData = bulkdata)
rm(bulkdata)

# Load single-cell Seurat object
sobj <- readRDS('sc_ref_pdac_seurat.rds')
gc()

#sobj
#colnames(sobj)
#nrow(sobj) #18507 features, 3000 variable features
#40k cells
#Features(sobj)
#Layers(sobj)

# Extract raw counts from Seurat object
sc_counts <- GetAssayData(sobj, layer = "counts")
gc()

# Retrieve and prepare single-cell metadata
cellmeta <- sobj@meta.data
celltype <- cellmeta[,c("cell_type", "patient")]
colnames(celltype) <- c("cellType", "SubjectName")
df <- data.frame(labelDescription=colnames(celltype), row.names = colnames(celltype))
rm(sobj)
gc()

# Create AnnotatedDataFrame object
sc_annot <- new("AnnotatedDataFrame", data=celltype, varMetadata=df)

# Create ExpressionSet object for single-cell reference
sc.eset <- ExpressionSet(assayData = as.matrix(sc_counts), phenoData = sc_annot) 
remove(cellmeta, celltype, sc_counts, sc_annot, df)
gc()

# Run Bisque decomposition (a.k.a. deconvolution)
res <- BisqueRNA::ReferenceBasedDecomposition(bulk.eset, sc.eset, markers=NULL, use.overlap=FALSE)
gc()

# Explore your results
bisque_results <- as.data.frame(res$bulk.props)
readr::write_csv(bisque_results, "bisque_results.csv")

```


## 4. CIBERSORTx 
Docker container also requires token from website. Go to the [CIBERSORTx website](https://cibersortx.stanford.edu/) and login with your account details. Get token.

```{r}

```


## 5. Immunedeconv 

```{r fig.width=8, fig.height=6, fig.align='center'}

#available methods
deconvolution_methods
#EPIC
?EPIC::EPIC.package

#run EPIC - input data should be a matrix
#tumour reference
EPIC_results <- EPIC(bulk = toydata, reference = TRef) #uses gene symbols

#results
EPIC_results$cellFractions
EPIC_results$fit.gof
EPIC_results_cf <- as.data.frame(EPIC_results$cellFractions)
fit <- as.data.frame(EPIC_results$fit.gof)
summary(EPIC_results_cf)
skimr::skim(EPIC_results_cf)

#export csv
write.table(EPIC_results_cf, "epic_results_demo.csv", sep = "\t") 

#transform data to plot 
results_gt <- EPIC_results_cf %>% rownames_to_column(var = "sample_id") %>% 
                                  gather(-sample_id, key = cell_type, value = proportion) 
glimpse(results_gt)
View(results_gt)

#reorder 
ordered <- results_gt %>% mutate(cell_type = fct_reorder(cell_type, proportion))

#plot 1 - dist boxplot 
ggplot(data = ordered, mapping = aes(x = cell_type, y = proportion, fill = cell_type)) + 
  geom_boxplot(outlier.alpha = 0, na.rm = TRUE) + 
  geom_point(size = 1.5, alpha = .5, position = position_jitter(seed = 1, width = .22)) + 
               labs(title = "EPIC estimates", 
                    x = "Cell type", y = "Proportion") + theme_minimal() + 
                    theme(axis.text.x = element_text(angle=90)) + 
                    scale_fill_brewer(palette = "Dark2") 

ggsave(file="epic_boxplot.png", height=7, width=8, dpi=400)


#plot 2 - stacked - distribution of cell types across samples
p <- ggplot(results_gt, aes(x = sample_id, y = proportion, fill = cell_type)) +
         geom_col(position = "fill") + theme_minimal() + 
          theme(axis.text.x = element_text(angle=90)) +  
           scale_fill_brewer(palette = "Dark2") +
           labs(title = "EPIC - Interactive plot", 
                    x = "Sample", 
                    y = "Estimated proportion") 
  
ggsave(file="epic_stacked.png", height=5, width=7, dpi=400)

library(plotly)
int <- ggplotly(p)
int

#save
htmlwidgets::saveWidget(as_widget(int), "interactive_plot.html")
```

## DREAM Challenge

##  Exercise 1- individual
```{r}

```

##  Exercise 2 - individual
```{r}

```

##  Exercise 3 - teams (breakout rooms)
```{r}

```



## References



Please visit the [**Deconvolution Course Repository**](https://github.com/estefvazquez/scDeconvolution_course) for further resources.  



- Chu, T., Wang, Z., Pe’er, D. Danko, C. G. Cell type and gene expression deconvolution with BayesPrism enables Bayesian integrative analysis across bulk and single-cell RNA sequencing in oncology. Nat Cancer 3, 505–517 (2022). https://doi.org/10.1038/s43018-022-00356-3  

- Wang, X., Park, J., Susztak, K., Zhang, N. R., & Li, M. (2019). Bulk tissue cell type deconvolution with multi-subject single-cell expression reference. Nature Communications, 10(1), 380. https://doi.org/10.1038/s41467-018-08023-x  

- Newman, A. M., Liu, C. L., Green, M. R., Gentles, A. J., Feng, W., Xu, Y., Hoang, C. D., Diehn, M., & Alizadeh, A. A. (2015). Robust enumeration of cell subsets from tissue expression profiles. Nature Methods, 12(5), 453–457. https://doi.org/10.1038/nmeth.3337  

- Jew, B., Alvarez, M., Rahmani, E., Miao, Z., Ko, A., Garske, K. M., Sul, J. H., Pietiläinen, K. H., Pajukanta, P., & Halperin, E. (2020). Publisher Correction: Accurate estimation of cell composition in bulk expression through robust integration of single-cell information. Nature Communications, 11(1), 2891. https://doi.org/10.1038/s41467-020-16607-9  

- Racle, J., de Jonge, K., Baumgaertner, P., Speiser, D. E., & Gfeller, D. (2017). Simultaneous enumeration of cancer and immune cell types from bulk tumor gene expression data. ELife, 6, e26476. https://doi.org/10.7554/eLife.26476  








